### PR TITLE
Delete method for Users

### PR DESCRIPTION
**Pull request by vlasovigor**
- [x] method

- [x] Test

- [x] exceptions
https://github.com/malinink/easy-tests/issues/363

### CODE REVIEW

**Comment by vlasovigor** [comment]
@malinink pls,look

**Comment by vlasovigor** [comment]
@malinink ready for review

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 1]
```diff
@@ -60,9 +62,33 @@ private Boolean isAdmin() {
...
-     */
+    @DeleteMapping("/{userId}")
+    public void delete(@PathVariable Integer userId) throws NotFoundException, ForbiddenException {
+        final UserModelInterface userModel = this.getUserModel(userId);
+
+        if (!this.isAdmin()) {
+            throw new ForbiddenException();
+        }
+
+        if (userModel == null) {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: already throw in getUserModel

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 2]
```diff
@@ -60,9 +62,33 @@ private Boolean isAdmin() {
...
+            throw new NotFoundException();
+        }
+
+        final UserModelInterface userModelForDelete = this.getUserModel(userId, this.usersOptionsBuilder.forDelete());
+        this.usersService.delete(userModelForDelete);
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: No, You should select for delete, and than delete with same option (for delete!)

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 3]
```diff
@@ -137,12 +138,60 @@ public void testListForbidden() throws Exception {
...
+        this.usersSupport.assertEquals(argumentCaptor.getValue(), userModel);
+    }
+
+    @Test
+    public void testDeleteForbidden() throws Exception {
+        final UserModelInterface userModel = new UserModel();
+        userModel.map(this.usersSupport.getEntityFixtureMock(0));
+        when(this.usersService.find(any(), any())).thenReturn(userModel);
+
+        final UserModelInterface userNotAdminModel = new UserModel();
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: may be it is a good idea to move mocking current user as admin and as not admin into separate functions in test.
i.e. withAdmin  withNotAdmin and after use that functions in all test methods.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by vlasovigor** [comment]
@malinink ready for review

**Comment by vlasovigor** [comment]
@malinink ready for review

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 4]
```diff
@@ -60,9 +61,21 @@ private Boolean isAdmin() {
...
+
+        if (!this.isAdmin()) {
+            throw new ForbiddenException();
+        }
+
+        if (userModel == null) {
+            throw new NotFoundException();
+        }
+
+        this.usersService.delete(userModel);
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: for delete option here

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 5]
```diff
@@ -137,12 +140,63 @@ public void testListForbidden() throws Exception {
...
+        final UserModelInterface userAdminModel = new UserModel();
+        userAdminModel.map(this.usersSupport.getAdminUser());
+        when(this.sessionService.getUserModel()).thenReturn(userAdminModel);
+
+        final UserModelInterface userModel = new UserModel();
+        userModel.map(this.usersSupport.getEntityFixtureMock(0));
+        when(this.usersOptionsBuilder.forDelete()).thenReturn(new UsersOptions());
+        when(this.usersService.find(any(Integer.class), any(UsersOptionsInterface.class))).thenReturn(userModel);
+
+        final ArgumentCaptor<UserModelInterface> argumentCaptor = ArgumentCaptor.forClass(UserModelInterface.class);
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: captor not need

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by vlasovigor** [comment]
@malinink pls,look

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 6]
```diff
@@ -73,12 +73,22 @@ public User show(@PathVariable Integer userId) throws NotFoundException, Forbidd
...
-    /**
-     * showMe
-     */
+
+    @DeleteMapping("/{userId}")
+    public void delete(@PathVariable Integer userId) throws NotFoundException, ForbiddenException {
+        final UsersOptionsInterface usersOptions = this.usersOptionsBuilder.forDelete();
+        final UserModelInterface userModel = this.usersService.find(userId, usersOptions);
+
+        if (!this.isAdmin()) {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: я бы вынес проверку isAdmin в самое начало, логично совсем ничего не делать, если не админ

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 7]
```diff
@@ -190,15 +191,63 @@ public void testShowNotFound() throws Exception {
...
-     */
+    @Test
+    public void testDeleteSuccess() throws Exception {
+        final UserModelInterface userAdminModel = new UserModel();
+        userAdminModel.map(this.usersSupport.getAdminUser());
+        when(this.sessionService.getUserModel()).thenReturn(userAdminModel);
+
+        final UserModelInterface userModel = new UserModel();
+        userModel.map(this.usersSupport.getEntityFixtureMock(0));
+        final UsersOptionsInterface usersOptionsForDelete=new UsersOptions();
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: откорректируйте код, нет пробелов

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 8]
```diff
@@ -190,15 +191,63 @@ public void testShowNotFound() throws Exception {
...
+        )
+                .andExpect(status().isOk())
+                .andExpect(content().string(""))
+                .andReturn();
+        verify(this.usersService, times(1)).delete(userModel,usersOptionsForDelete);
+    }
+
+    @Test
+    public void testDeleteForbidden() throws Exception {
+        final UserModelInterface userModel = new UserModel();
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: в тесте, после изменений метода и переноса isAdmin в верх, можно будет убрать половину моков.

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 9]
```diff
@@ -190,15 +191,63 @@ public void testShowNotFound() throws Exception {
...
+                .andReturn();
+    }
+
+    @Test
+    public void testDeleteNotFound() throws Exception {
+        final UserModelInterface userAdminModel = new UserModel();
+        userAdminModel.map(this.usersSupport.getAdminUser());
+        when(this.sessionService.getUserModel()).thenReturn(userAdminModel);
+
+        final UsersOptionsInterface usersOptionsForDelete=new UsersOptions();
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: cs по пробелам

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by vlasovigor** [comment]
@malinink pls,look again

**Review by malinink — APPROVED** [review]

**Pull request merged**